heroku_connect/models.py

```diff
@@ -0,0 +1,274 @@
+
+
+class TriggerLogQuerySet(models.QuerySet):
+    """A QuerySet for trigger log models."""
+
+    def failed(self):
+        """Filter for log records with sync failures."""
+        return self.filter(state=TRIGGER_LOG_STATE['FAILED'])
+
+    def related_to(self, instance):
+        """Filter for all log objects of the same connected model as the given instance."""
+        return self.filter(table_name=instance.table_name, record_id=instance.record_id)
+
+
+TRIGGER_LOG_ACTION = {
+    'INSERT': 'INSERT',
+    'UPDATE': 'UPDATE',
+    'DELETE': 'DELETE',
+}
+"""The type of change that a trigger log object represents."""
+
+
+TRIGGER_LOG_ACTION_CHOICES = sorted((value, value) for value in TRIGGER_LOG_ACTION.values())
+
+
+TRIGGER_LOG_STATE = {
+    'SUCCESS': 'SUCCESS',
+    'MERGED': 'MERGED',
+    'IGNORED': 'IGNORED',
+    'FAILED': 'FAILED',
+    'READONLY': 'READONLY',
+    'NEW': 'NEW',
+    'IGNORE': 'IGNORE',
+    'PENDING': 'PENDING',
+    'REQUEUE': 'REQUEUE',
+    'REQUEUED': 'REQUEUED',
+}
+"""The sync state of the change tracked by a trigger log entry."""
+
+
+TRIGGER_LOG_STATE_CHOICES = sorted((value, value) for value in TRIGGER_LOG_STATE.values())
+
+
+class TriggerLogAbstract(models.Model):
+    """
+    Support for accessing the Heroku Connect Trigger Log data and related actions.
+
+    Heroku Connect uses a Trigger Log table to track local changes to connected models (that is,
+    in the Heroku database. Such changes are recorded as rows in the trigger log and, for
+    read-write mappings, eventually written back to Salesforce.
+
+    Old logs are moved to an archive table (usually after 24 hours), from where they are purged
+    eventually (currently 30 days for paid plans, 7 days for demo). Recent logs are modeled by
+    :class:`TriggerLog`; archived logs by :class:`TriggerLogArchive`.
+
+    The data represented by these models is maintained entirely by Heroku Connect, and is
+    instrumental to its operations; it should therefore not be modified. A possible exception is
+    the ``state`` field, which may be changed as detailed in the `error handling`_ section in the
+    Heroku Connect documentation.
+
+    .. seealso::
+        - :class:`.TriggerLogQuerySet`
+        - `Trigger Log in Heroku Connect docs`_
+
+    .. _Trigger Log in Heroku Connect docs:
+        https://devcenter.heroku.com/articles/writing-data-to-salesforce-with-heroku-connect#understanding-the-trigger-log
+    .. _error handling:
+        https://devcenter.heroku.com/articles/writing-data-to-salesforce-with-heroku-connect#write-errors
+
+    """
+
+    # I18N / TRANSLATIONS:
+    #
+    # Field names and choices don't use translations, because they refer to (English)technical
+    # terms in Heroku Connect's manual. That connection should be preserved. Trigger Log Models
+    # are for technical folks and not intended to be user-facing.
+
+    # read-only fields
+    id = models.BigIntegerField(primary_key=True, editable=False)
```

Comment:
Isn't this a serial, meaning it has a default `nextval(sequence)`.
I believe those are positive integers, but I would use the AutoField, which is the default. So I would just drop the line.

Reply:
I see this as explicit documentation of the unmanaged table: yes, this column exists, and it's some form of big integer PK. We're not managing it, so it might have autofield behavior, or it might not. I could make it an `AutoField` for easier testing, but I'd really like to have it there explicitly.

Reply:
Please use the `AutoField` than. I check it does have `nextval('salesforce._trigger_log_id_seq'::regclass)` as a default, so is a postgres `serial` not `int`.

Reply:
We don't have it for the BaseModel ether. If you want to overwrite the default go for it, I don't mind ether way.

Reply:
Github is showing old code here, because I already changed the line. It's currently a `BigAutoField`.